initial version of the adapter

demmer opened the pull request:
Initial version of the google analytics adapter.

I still need to set up travis but the code is all there and the tests have almost 100% coverage when run locally.

demmer committed:
initial version of the adapter

demmer commented:
@mstemm @davidvgalbraith Please take a look at the code.
@dmehra @rlgomes Please take a look at the README and travis setup. Also try it for yourself :)

demmer committed:
add sort to fix reduce -every across properties / views

demmer committed:
minor fixes and cleanup to get coverage to 100%

* Fix the test and the sorting code to handle missing result rows.
* Add a test that it fails when given a filter.
* Fix up the logic for extracting the field name to not have any
  unreachable branches.

mstemm commented on README.md:
````diff
@@ -1,2 +1,159 @@
+```bash
+$ npm install juttle-engine
+```
+
````
Can you add the ecosystem section here to match what the other adapters have?

mstemm commented on README.md:
```diff
@@ -1,2 +1,159 @@
+
+### Read options
+
+Since google analytics returns aggregate data, the only supported query pattern using Juttle uses a combination of [read](http://juttle.github.io/juttle/processors/read/) and [reduce](http://juttle.github.io/juttle/processors/reduce/), using the general form:
```
This was a bit vague for me. I think you're saying that ga only provides summary data and not raw data, making a subsequent reduce the best way to map the summary back into the read?

mstemm commented on lib/google.js:
```diff
@@ -0,0 +1,222 @@
+let auth, authorized, accountId;
+
+function init(config) {
+    auth = new google.auth.JWT(
```
At this point can you check that the config has a 'service_account' property and 'client_email'/'private_key' subproperties? I accidently used 'service-account' and the error was pretty ugly.

mstemm commented:
I'd still like to go through the install process myself but ran into a problem I'll work out with demmer tomorrow. I think somebody else with more familiarity with optimizers should look at the optimizer, but the rest looked good.

Also I found the name 'ga' confusing--I was wondering what the general availability adapter was. I know 'googleanalytics' is too long. How about 'ganalytics'?

mstemm commented:
The problem I ran into was that I didn't have permissions to add the service account to the google analytics account. Once @demmer gave me permissions, the rest worked.

davidvgalbraith commented:
Agree with @mstemm that `read ga` sounds like "goo goo ga ga". I'd do the full `read google_analytics`, as typing is less of a burden than we sometimes make it out to be.

davidvgalbraith commented on README.md:
````diff
@@ -1,2 +1,159 @@
+```
+
+```
+┌──────────────┬────────────┬─────────────────────────────────────────┐
````
This table is pretty ghetto, how about something like https://github.com/juttle/juttle-elastic-adapter/pull/98/files#diff-04c6e90faac2675aa89e2176d2eec7d8R123?

davidvgalbraith commented on lib/read.js:
```diff
@@ -0,0 +1,84 @@
+
+        // to provide better error messages, the optimizer includes information
+        // as to the invalid invocation in the optimization_info
+        if (params.optimization_info.invalid) {
```
Can we fail a little faster by throwing this from the optimizer?

demmer replied:
Good idea. See ecaa1444ed32218c06dc3a73983772296c9597ff.

davidvgalbraith commented on lib/google.js:
```diff
@@ -0,0 +1,222 @@
+            } else if (spec.dataType === 'TIME') {
+                value = parseFloat(value);
+            }
+            switch (name) {
```
What's going on here? Seems like some of these set the time and others add something to it so if they come in the wrong order we'll have some problems.

demmer replied:
Yep, that's what it does. But google always gives back the dimensions in the order we requested them, so since we always put the moment part before the offset adjustment it works out.

davidvgalbraith replied:
Seems like a house of cards, can we put a comment outlining that expectation?

demmer replied:
Done.

davidvgalbraith commented on lib/optimize.js:
```diff
@@ -0,0 +1,156 @@
+    if (node.type === 'UnaryExpression' && node.operator === '*' && node.argument.d) {
+        return node.argument.value;
+
+    } else { // node.type === 'Field'
```
Can we assert this instead of blindly trusting this comment?

demmer replied:
I originally did that. But then my code coverage was < 100% so I restructured this way.

Admittedly that's a bit cowardly, but so is adding a branch to throw an Error (that I can't ever trip) and then reducing the branch coverage to something like 99.X%. So I went this route instead.

davidvgalbraith commented on lib/optimize.js:
```diff
@@ -0,0 +1,156 @@
+            }
+
+            dimensions = dimensions.concat(_.difference(groupby, META_GROUPBY));
+            if (_.contains(groupby, 'view') || _.contains(groupby, 'viewId')) {
```
Looks like if I group by `view` and `webProperty` I only get `view`?

demmer replied:
Not exactly... the `metagroup` ends up affecting a couple things in the query:
1. If the metagroup is `view`, then all views are included in the results, not just the "All Web Site Data" view.
2. If the metagroup is set at all, then we tack on the full view metadata (including view, viewId, webProperty, and webPropertyId) into each point.

In either case, the specific groupby keys that the user selected are then handled by the native juttle `reduce` that's left in the flowgraph since optimization always returns false.

dmehra commented:
I'm having trouble with `-from/-to/-last` when doing `read ga`. For example, `read ga -last :52 weeks: | reduce -every :day:` gives me data for 2015-03-24 to 2015-04-27, and `read ga -last :year: | reduce -every :week:` gave me data for just one day of 2015-01-08.

demmer committed:
google: fix time calculations for reduce -every :week:

demmer committed:
README: updates in response to PR feedback

demmer committed:
fix error handling when reducing without a metric

demmer committed:
add support for paging through > 1000 results

demmer committed:
read: require -from and -to to be aligned to whole days

Since the underlying API only accepts from and to intervals of whole days,
add corresponding restrictins in the implementation.

demmer committed:
rename to "googleanalytics" from "ga"

demmer commented:
@dmehra @davidvgalbraith @mstemm I pushed a bunch of commits that include fixes for the time bounds, adding support for paging, and renaming the adapter to `googleanalytics`. Please look again but then I'll publish.

demmer committed:
optimize: fix code to properly check for unexpected node types

Add the appropriate comment so istanbul ignores the unreachable
branch for the purposes of code coverage.

davidvgalbraith commented on lib/read.js:
```diff
@@ -0,0 +1,113 @@
+        this.fetchSize = options.fetchSize;
+
+        if (this.options.from) {
+            if (! this.options.from.quantize(JuttleMoment.duration(1, 'd')).eq(this.options.from)) {
```
This is a pretty involved `if`, can we get a little `function isRoundDay(moment)` that we'd call here and on line 61?

demmer replied:
Good idea. Added.

davidvgalbraith commented on lib/read.js:
```diff
@@ -0,0 +1,113 @@
+        if (to.eq(this.now)) {
+            endDate = JuttleMoment.format(to, 'YYYY-MM-DD');
+        } else {
+            endDate = JuttleMoment.format(to.subtract(JuttleMoment.duration(1, 'd')), 'YYYY-MM-DD');
```
What's going on here?

demmer replied:
Added a comment to try to explain it. Make sense now?

demmer committed:
read: add isRoundDay helper function

demmer committed:
read: add comment explaining endDate calculation

davidvgalbraith commented:
+1

demmer committed:
optimize: move the error for invalid optimizations into the optimizer

Rather than relying on read to throw the error if the optimization
was invalid, throw the error from the optimizer itself.
